### PR TITLE
Add Warp recipe

### PR DESCRIPTION
**Pull request by alisterburt**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by alisterburt** [comment]
@dtegunov please comment here to be added as a maintainer of this recipe

File: recipes/warp/build.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+cd ${PROJECT_ROOT}
+
+# build LibTorchSharp
+echo building LibTorchSharp
+cd LibTorchSharp
+mkdir build
+cd build
+TORCH_CMAKE_PREFIX=`python -c 'import torch;print(torch.utils.cmake_prefix_path)'`
+CUSTOM_CMAKE_ARGS="-DCMAKE_PREFIX_PATH="+${TORCH_CMAKE_PREFIX}
+cmake ${CMAKE_ARGS} ${CUSTOM_CMAKE_ARGS} ..
```
Comment (alisterburt) [inline recipes/warp/build.sh:20]:
```suggestion
echo ${CUSTOM_CMAKE_ARGS} ${CMAKE_ARGS}
cmake ${CUSTOM_CMAKE_ARGS} ${CMAKE_ARGS}  ..
cmake ${CMAKE_ARGS} ${CUSTOM_CMAKE_ARGS} ..
```

File: recipes/warp/build.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+make
+cd ${PROJECT_ROOT}
+
+# build LibTorchSharp
+echo building LibTorchSharp
+cd LibTorchSharp
+mkdir build
+cd build
+TORCH_CMAKE_PREFIX=`python -c 'import torch;print(torch.utils.cmake_prefix_path)'`
+CUSTOM_CMAKE_ARGS="-DCMAKE_PREFIX_PATH="+${TORCH_CMAKE_PREFIX}
```
Comment (alisterburt) [inline recipes/warp/build.sh:19]:
```suggestion
CUSTOM_CMAKE_ARGS="-DCMAKE_PREFIX_PATH=${TORCH_CMAKE_PREFIX}"
```

**Comment by dtegunov** [comment]
Confirmed

File: recipes/warp/build.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+cp ${DOTNET_DLL_DIR}/Noise2Half.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/Noise2Class.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/EstimateWeights.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/Frankenmap.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/MrcConverter.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/WarpWorker.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/WarpTools.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/MTools.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/MCore.dll ${PREFIX}/lib/
+cp ${DOTNET_DLL_DIR}/Snippets.dll ${PREFIX}/lib/
```
Comment (alisterburt) [inline recipes/warp/build.sh:71]:
```suggestion
echo "listing contents of ${DOTNET_DLL_DIR}"
ls ${DOTNET_DLL_DIR}
echo "---"
cp ${DOTNET_DLL_DIR}/Noise2Map.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Noise2Mic.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Noise2Tomo.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Noise2Half.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Noise2Class.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/EstimateWeights.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Frankenmap.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/MrcConverter.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/WarpWorker.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/WarpTools.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/MTools.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/MCore.dll ${PREFIX}/lib/
cp ${DOTNET_DLL_DIR}/Snippets.dll ${PREFIX}/lib/
```

File: recipes/warp/build.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
```
Comment (alisterburt) [inline recipes/warp/build.sh:74]:
```suggestion
for file in **.dll; do
    echo ${file}
    cp $file ${PREFIX}/lib/
```

File: recipes/warp/build.sh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+cp ${DOTNET_PUBLISH_DIR}/WarpWorker ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/WarpTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MCore ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/Snippets ${PREFIX}/bin/
+
+# copy dlls into the lib directory in the conda environment
+echo "listing contents of ${DOTNET_DLL_DIR}"
+ls ${DOTNET_DLL_DIR}
+echo "---"
```
Comment (alisterburt) [inline recipes/warp/build.sh:61]:
```suggestion
echo "copying dotnet generated dlls into ${PREFIX}/lib"
```

File: recipes/warp/build.sh [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+for file in **.dll; do
+    echo ${file}
```
Comment (alisterburt) [inline recipes/warp/build.sh:63]:
```suggestion
```

File: recipes/warp/build.sh [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    cp $file ${PREFIX}/lib/
```
Comment (alisterburt) [inline recipes/warp/build.sh:62]:
```suggestion
for file in $(find /path/to/directory -type f -name "*.dll"); do
    cp -v ${file} ${PREFIX}/lib/
done
```

File: recipes/warp/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    - cudnn
+  run:
+    - dotnet ==6.0
+    - pytorch ==2.1.0
+    - libtorch ==2.1.0
+    - fftw
+    - libtiff
+    - zlib
+    - torchvision
+    - cudnn
```
Comment (dhirschfeld) [inline recipes/warp/meta.yaml:47]: Pinning to patch versions is a anti-pattern that means you won't pick up any security patches and nearly guarantees that your package will have problems resolving.
```suggestion
  build:
    - {{ compiler('cxx') }}
    - {{ compiler('cuda') }}
    - cmake
    - dotnet ~=6.0
    - pytorch ~=2.1.0
    - libtorch ~=2.1.0
    - fftw
    - libtiff
    - zlib
    - cudnn
    - python
  host:
    - dotnet ~=6.0
    - pytorch ~=2.1.0
    - libtorch ~=2.1.0
    - fftw
    - libtiff
    - zlib
    - torchvision
    - cudnn
  run:
    - dotnet ~=6.0
    - pytorch ~=2.1.0
    - libtorch ~=2.1.0
    - fftw
    - libtiff
    - zlib
    - torchvision
    - cudnn
```
Using https://peps.python.org/pep-0440/#compatible-release

**Comment by dhirschfeld** [comment]
```
Solving environment (_test_env): ...working... failed
WARNING: failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform linux-64: {MatchSpec("warp==2.0.0dev0=h6efa04d_0"), MatchSpec("6.0")}
Encountered problems while solving:
  - nothing provides dotnet 6.0 needed by warp-2.0.0dev0-h6efa04d_0

Could not solve for environment specs
The following package could not be installed
└─ warp 2.0.0dev0 h6efa04d_0 is not installable because it requires
   └─ dotnet 6.0 , which does not exist (perhaps a missing channel).
```

**Comment by dhirschfeld** [comment]
image
https://anaconda.org/conda-forge/dotnet/files?page=2

**Comment by alisterburt** [comment]
thanks for the input @dhirschfeld, very much appreciated! I hadn't come across the `~=` operator for versioning

Your comment is hidden, copying in case it is useful to others: 

> Pinning to patch versions is a anti-pattern that means you won't pick up any security patches and nearly guarantees that your package will have problems resolving. https://peps.python.org/pep-0440/#compatible-release

**Comment by dhirschfeld** [comment]
> *`Merge branch 'main' into warp`*

I'm not sure what's going on there, but in general you don't need to merge `main` into a `staged-recipes` PR. Since the error seems so unusual I'd be inclined to rebase on the latest main:
```
git switch main
git pull
git switch warp
git rebase main
git push -f
```

**Comment by alisterburt** [comment]
@dhirschfeld there was a 'sync with main' prompt on GitHub so I clicked it, agree it is a bit weird in this case and shouldn't be necessary/have changed anything

**Comment by dhirschfeld** [comment]
Before the merge it looks like you had a green check mark so perhaps just revert to that commit

File: recipes/warp/bld.bat [hunk 9]
```diff
@@ -0,0 +1,72 @@
+@echo off
+setlocal
+
+REM there is no PyTorch>=2 available on conda-forge, forceinstall from pytorch channel
+conda install pytorch::pytorch --channel pytorch~=2.1
```
Comment (alisterburt) [inline recipes/warp/bld.bat:5]:
```suggestion
conda install pytorch pytorch-cuda=11.8 -c pytorch -c nvidia
```

File: recipes/warp/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  skip: true  # [osx]
+  number: 0
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}
+    - cmake
+    - dotnet ~=6.0
+    - pytorch ~=2.1[build=cuda*]  # [not win]
```
Comment (alisterburt) [inline recipes/warp/meta.yaml:23]:
```suggestion
    - pytorch ~=2.1  # [not win]
```

File: recipes/warp/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    - fftw
+    - libtiff
+    - zlib
+    - cudnn
+    - python
+  host:
+    - dotnet ~=6.0
+    - pytorch ~=2.1[build=cuda*]  # [not win]
```
Comment (alisterburt) [inline recipes/warp/meta.yaml:31]:
```suggestion
    - pytorch ~=2.1  # [not win]
```

File: recipes/warp/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    - fftw
+    - libtiff
+    - zlib
+    - torchvision
+    - cudnn
+  run:
+    - dotnet ~=6.0
+    - pytorch ~=2.1[build=cuda*]  # [not win]
```
Comment (alisterburt) [inline recipes/warp/meta.yaml:39]:
```suggestion
    - pytorch ~=2.1  # [not win]
```

File: recipes/warp/build.sh [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+cp ${DOTNET_PUBLISH_DIR}/MrcConverter ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/WarpWorker ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/WarpTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MCore ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/Snippets ${PREFIX}/bin/
+
+# copy dlls into the lib directory in the conda environment
+echo "copying dotnet generated dlls into ${PREFIX}/lib"
+for file in $(find /path/to/directory -type f -name "*.dll"); do
```
Comment (alisterburt) [inline recipes/warp/build.sh:59]:
```suggestion
for file in $(find Release -type f -name "*.dll"); do
```

File: recipes/warp/build.sh [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+cp ${DOTNET_PUBLISH_DIR}/MrcConverter ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/WarpWorker ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/WarpTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MTools ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/MCore ${PREFIX}/bin/
+cp ${DOTNET_PUBLISH_DIR}/Snippets ${PREFIX}/bin/
+
+# copy dlls into the lib directory in the conda environment
+echo "copying dotnet generated dlls into ${PREFIX}/lib"
+for file in $(find Release -type f -name "*.dll"); do
```
Comment (alisterburt) [inline recipes/warp/build.sh:59]:
```suggestion
for file in ${DOTNET_DLL_DIR}/*.dll; do
```

**Comment by dhirschfeld** [comment]
You may need a rerender after changing the skip, though not entirely sure how that works in `staged-recipes`. 

Sometimes it's easier to get a minimal working version published and to debug issues in PRs in the actual feedstock.

**Comment by alisterburt** [comment]
Thanks @dhirschfeld, I've pushed changes and we should have a minimal passing build now - a few things are missing from the package checklist provided but if you're happy to merge this as is I can sort those things out in PRs to the feedstock?

**Comment by dhirschfeld** [comment]
I can't actually merge. Once CI is green you can ping an appropriate team and hopefully someone with merge rights can do the honours.

https://github.com/conda-forge/staged-recipes#12-my-pull-request-passes-all-checks-but-hasnt-received-any-attention-how-do-i-call-attention-to-my-pr--what-is-the-customary-amount-of-time-to-wait

**Comment by alisterburt** [comment]
Asked for help in gitter: copying here for clarity

I have a package which is failing during "Run docker build" with a strange error about an empty version string:
```
rendering /home/conda/staged-recipes-copy/recipes/warp for linux-64
WARNING: Setting build platform. This is only useful when pretending to be on another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/conda/models/version.py", line 43, in __call__
    return cls._cache_[arg]
KeyError: ''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 261, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 151, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 188, in build_folders
    G = construct_graph(recipes_dir, worker=worker, run='build',
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 420, in construct_graph
    add_recipe_to_graph(recipe_dir, graph, run, worker, conda_resolve,
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 254, in add_recipe_to_graph
    add_dependency_nodes_and_edges(name, graph, run, worker, conda_resolve, config=config,
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 487, in add_dependency_nodes_and_edges
    if not _installable(dep, version, build_str, metadata.config, conda_resolve):
  File "/home/conda/staged-recipes-copy/.ci_support/compute_build_graph.py", line 440, in _installable
    installable = conda_resolve.find_matches(ms)
  File "/opt/conda/lib/python3.10/site-packages/conda/resolve.py", line 918, in find_matches
    res = tuple(p for p in candidate_precs if spec.match(p))
  File "/opt/conda/lib/python3.10/site-packages/conda/resolve.py", line 918, in <genexpr>
    res = tuple(p for p in candidate_precs if spec.match(p))
  File "/opt/conda/lib/python3.10/site-packages/conda/models/match_spec.py", line 255, in match
    if not self._match_individual(rec, field_name, v):
  File "/opt/conda/lib/python3.10/site-packages/conda/models/match_spec.py", line 262, in _match_individual
```

The recipe contains the following version string, but I'm not sure this is the version string being referenced in the build step.

```
{% set name = "Warp" %}
{% set version = "2.0.0dev0" %}

package:
  name: {{ name|lower }}
  version: {{ version }}
```

There is a similar error reported in conda-build from 2016, it was apparently caused by some invalid state and was fixed by removing $CONDA_ROOT/conda-bld https://github.com/conda/conda-build/issues/1197

A previous build attempt from the same source and with the same version number did not suffer the same error, this makes me think it is not a problem with my version number. https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=915786&view=logs&jobId=eb6ecf2b-0363-5c1a-baa7-2d771889d197&j=eb6ecf2b-0363-5c1a-baa7-2d771889d197&t=248aa990-2a52-51cf-fcc8-1e0aefce9a33

Does anyone have any ideas what might be the issue and how to solve it?

**Comment by alisterburt** [comment]
closing as we had problems building against the conda-forge distribution of pytorch and moved to using conda-build in our own infra and deploying to another channel

**Pull request closed without merging**